### ISSUE TITLE
maxNumberOutputs bug

### ISSUE THREAD

**Issue by mlucool** [jupyterlab/jupyterlab#10045]
## Description

The outputs after the first N/2 can sometimes be wrong with `maxNumberOutputs: 50` set.
## Reproduce

```
import random
for i in range(100):
    for x in range(random.randint(1, 100)):
        print(x, flush=True)
    for x in range(random.randint(1, 10)):
        print(x, file=sys.stderr, flush=True)
```
This only shows the first line:
![image](https://user-images.githubusercontent.com/1813603/113628373-cdc98780-9632-11eb-8585-ca256543feac.png)

The cell outputs look correct when viewing in the JSON editor:
![image](https://user-images.githubusercontent.com/1813603/113628723-50524700-9633-11eb-8e30-894f3d4bb287.png)

## Expected behavior

The whole stderr/stdout is rendered

## Context

- Operating System and version: Windows
- Browser and version: Chrome 
- JupyterLab version: `2.3.0rc0`

CC @echarles @goanpeca

**Comment by mlucool** [comment]
The randomness isn't helpful here. This should be easier to reason about:

```
for i in range(100):
    for x in range(10):
        print(x, flush=True)
    for x in range(10):
        print(x, file=sys.stderr, flush=True)
```

**Comment by echarles** [comment]
As discussed with @goanpeca, I am working on this.

**Comment by echarles** [comment]
I have debugged this issue and the behavior, although not good, can be explained (with my current understanding) with:

Step 1: If we have more divs than we have defined, the displayed widget (which corresponds to a div) is removed.

```js
    } else if (index >= this.maxNumberOutputs) {
      console.log('---', output)
      layout.removeWidgetAt(this.headTailNumberOutputs + 1);
      layout.insertWidget(index, output);
    }
```

In the given reproducer, the kernel has not yet finished populating its div, so at a later point, a kernel message comes in and is aimed to be displayed with 

```js
 /**
   * Update an output in the layout in place.
   */
  private _setOutput(index: number, model: IOutputModel): void {
    console.log('---', index, model);
    const layout = this.layout as PanelLayout;
    const panel = layout.widgets[index] as Panel;
```

... which fails with an exception as the widgets[index] is undefined.

Option 1: hides instead of remove, which will not help the goal of the stripped outputs which is to have better performance for outputs with many divs.

```js
    } else if (index >= this.maxNumberOutputs) {
      console.log('---', output)
      (layout.widgets[this.headTailNumberOutputs + 1] as Panel).addClass('jp-Output-Hidden';
      layout.insertWidget(index, output);
    }
```

Option 2: Maintain a copy of the layout (and its widgets array): the memory mode, and the effective rendered dom. This is a huge change and is not desirable IMHO.

Any other options?

**Comment by mlucool** [comment]
This is causing confusion for some users as data looks wrong at times because it is partially cutoff with no clear indication of such. Wrong data is worse than a slow experience.

I am not familiar with this code, but some ideas (many of which may be not possible):
1. Treat text as specials and do something more bespoke (maybe it'll solve https://github.com/jupyterlab/jupyterlab/issues/10046). See also [LimitOutput](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/limit_output/readme.html).
2. Teach layout that some widgets should not be rendered
3. Test if wrapping the DOM that should be hidden with a single parent node would get us far enough along with performance (so the dom is added, but a single check removes it from most calculations).

**Comment by echarles** [comment]
https://github.com/jupyterlab/jupyterlab/pull/11509 is merged in master (targets 3.3).

I have the fix backported on 3.2.x with https://github.com/jupyterlab/jupyterlab/pull/11509 (still needs to be merged).
